crates/fluvio/src/producer/memory_batch.rs

```diff
@@ -8,6 +8,39 @@ use fluvio_types::Timestamp;
 
 use super::*;
 
+pub type BatchSize = usize;
+pub enum BatchRecordStatus {
+    DontFit(BatchSize, Record), // Actual record does not fit in the batch
+    Full(Offset),               // Actual record fits in the batch and the batch is full.
+    NotFull(Offset),            // Actual record fits in the batch and the batch is not full.
+}
+
+impl BatchRecordStatus {
+    pub fn fit(&self) -> bool {
+        match self {
+            BatchRecordStatus::DontFit(_, _) => false,
+            BatchRecordStatus::Full(_) => true,
+            BatchRecordStatus::NotFull(_) => true,
+        }
+    }
+
+    pub fn is_full(&self) -> bool {
+        match self {
+            BatchRecordStatus::DontFit(_, _) => false,
+            BatchRecordStatus::Full(_) => true,
+            BatchRecordStatus::NotFull(_) => false,
+        }
+    }
+
+    pub fn not_full(&self) -> bool {
+        match self {
+            BatchRecordStatus::DontFit(_, _) => false,
+            BatchRecordStatus::Full(_) => false,
+            BatchRecordStatus::NotFull(_) => true,
+        }
+    }
```

Comment:
I think this is a little confusing too.  Full/not_full is already a property of the state of the Memory Batch, having this encoded into a temporary result I think ends up duplicating that state in a confusing way.

So I this the memory batch return from push_limit could also use a result type Ok (for full and not full),  and returning an error with (est_size, record).  This would also make the push_records at both levels consistent.

Reply:
Sure! I added one commit using result instead of a new enum